book.md

```diff
@@ -959,7 +959,110 @@ code pointers: attacks against non-control data, which will be the topic of
+Since these types of attacks are prominent, many mitigations have been designed
+with the goal of maintaining control-flow integrity. Non-control data
+attacks\index{non-control data attacks}, also known as data-only
+attacks\index{data-only attacks}, can completely bypass these mitigations,
+since the data they modify is not the control data that these mitigations
+protect.
+
+Non-control data attacks can range from very simple attacks targeting a single
+piece of data to very elaborate attacks with very high expressiveness. A very
+simple example may look something like this:
+
+```
+// Returns zero for failure, non-zero for success.
+int authenticate() {
+  int authenticated = 0;
+  char passphrase[10];
+  if (fgets(passphrase, 20, stdin)) {     // buffer overflow
+    if (!strcmp(passphrase, "secret\n")) {
+      authenticated = 1;
+    }
+  }
+  return authenticated;
+}
+```
+
+The example shows a simplified[^non-control-example] function that reads a
+passphrase from a user, compares it with a known value and sets an integer
+stack variable to indicate whether "authentication" was successful or not. The
+function contains a very obvious buffer overflow, as the string length limit
+passed to `fgets` does not match the buffer size.
+
+[^non-control-example]: This is obviously not a realistic example of how
+  authentication should be done, but simply serves to illustrate how a
+  buffer overflow into a non-control variable can have serious security
+  consequences.
+
+Figure @fig:non-control-data-attack shows the stack frame layout for this
+function when the code is compiled for AArch64 with Clang 10.0[^ymmv]. As the figure
+shows, an overflow of `passphrase` will overwrite `authenticated`, setting it
+to a non-zero value, even though the passphrase was incorrect.  The
+`authenticate` function will then return a non-zero value, incorrectly
+indication authentication success.
+
+[^ymmv]: The stack frame layout may be significantly different for other
+  architectures and compilers.
+
+![Stack frame for `authenticate`](img/non-control-data-attack){ width=60% #fig:non-control-data-attack }
+
+For many more simple examples of data-only attacks that can occur in real
+applications, see [@Chen2005]. Although this makes it clear that data-only
+attacks are a real issue, it leaves open a very important question: what are
+the limits of such attacks? It is tempting to assume that data-only attacks are
+somehow inherently limited, however it has been demonstrated in [@Hu2016] that
+they can, in fact, be very expressive[^caveat].
+
+[^caveat]: The authors describe how DOP gadgets can be chained to simulate
+  a Turing machine, making DOP attacks Turing-complete (it's not possible
+  to simulate the infinite tape of a Turing machine on any actual hardware,
+  of course). Turing-completeness is not, however, a particularly useful
+  measure of exploitability, as explained in [@Dullien2018].
```

Comment:
I wonder if the reason why Turing-completeness is not a useful measure of exploitability could be summarized here or elsewhere in the book? It seems like something interesting to know for most readers of the book?

Reply:
I'm adding a sentence to explain this briefly.